### PR TITLE
Gangplank skills

### PR DESCRIPTION
**Pull request by xTeJk**

### CODE REVIEW

**Comment by MingSaber976** [comment]
GangPlankE you can use GameScript to addParticle(if you know what to do),and about give buffs to ally,you can follow "Taric R" to make AOE buff.

**Comment by MingSaber976** [comment]
GangPlank W should be scale by AP, not AD.
We don't have "RestoreHealth" function, please fix it

**Comment by MingSaber976** [comment]
About R spell, you can use `for()` to make codes shorter.

**Comment by MingSaber976** [comment]
Gangplank W you can use
`owner.Stats.CurrentHealth += healing number;` to restore health.

**Comment by MingSaber976** [comment]
GangplabkE BuffHudVisual and Particle should be in `GameScript`, else it will be weird, like particles didn't remove on other people.

**Comment by MingSaber976** [comment]
in GangplankW, that Game script have "removeAfter" things, you could add that and remove `owner.RemoveGameScript();` function.

**Comment by xTeJk** [comment]
> GangplabkE BuffHudVisual and Particle should be in `GameScript`, else it will be weird, like particles didn't remove on other people.

AddBuffHUDVisual should be defined as var and then removed by RemoveBuffHUDVisual(..); in timer?

**Comment by MingSaber976** [comment]
But you use `foreach()`, mean all units in the range.
But you coded as removeParticle on one's(not sure who is the one).

**Comment by MingSaber976** [comment]
So coded in GameScript then it won't have confusion and it will be more readable.

**Comment by MingSaber976** [comment]
Check the `IGameScript`( not the buff one)construction is the current version, because you don't have "IChampion, ISpell,IAttackableUnit" type ( all without the "I" word)

**Comment by MingSaber976** [comment]
CreateTimer should be vars number, not 0.5f (else it will happen at the same time)

**Comment by xTeJk** [comment]
if (unit is IChampion || unit is Minion || unit is Monster)
Im not sure about this construction (shouldnt be IMinion and IMonster?)

About E foreach particle removing also.

I've checked GameScript but still dont understand how removeAfter; is working and what i should paste in code

**Comment by xTeJk** [comment]
> CreateTimer should be vars number, not 0.5f (else it will happen at the same time)

So it should be CreateTimer(0.5, () => instead of CreateTimer(0.5f, () => am i right?

**Comment by MingSaber976** [comment]
Gangplank visualHud and Particles should be in "Buffs/GangplankE.cs", so after time, it will disappear on units(remember foreach() function, you can use that)

**Comment by MingSaber976** [comment]
@xTeJk No, like this example.
```
for(float p = 0.0f; p<=12; p += 0.5f)
{
  CreateTimer(p,()=>
  {
    ............
  });
}
```

**Comment by MingSaber976** [comment]
And test in LeagueSandbox make sure it all works or not, most coding problems were solved.

**Comment by MingSaber976** [comment]
And please remove GangplankR because the spells construction not that perfect(especially CreateTimer() function,the developers don't recommended to use that)

**Comment by xTeJk** [comment]
> And please remove GangplankR because the spells construction not that perfect(especially CreateTimer() function,the developers don't recommended to use that)

Sure.



> And test in LeagueSandbox make sure it all works or not, most coding problems were solved.

Everything is working fine.

**Comment by MingSaber976** [comment]
GangplankE have `CreateTimer` please remove it and if you want removeParticle, code them in the BuffGameScript(Buffs/GangplankE.cs)

File: Buffs/GangplankE/GangplankE.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            _statMod.MoveSpeed.PercentBonus = _statMod.MoveSpeed.PercentBonus + (10f + 5f * ownerSpell.Level) / 100f;
+            _statMod.AttackDamage.PercentBonus = _statMod.AttackDamage.PercentBonus + (10f + 10f * ownerSpell.Level) / 100f;
+            unit.AddStatModifier(_statMod);
+            
+            var time = 7.0f;
+            _hudvisual = AddBuffHUDVisual("RaiseMorale", time, 1, unit);
+            
+            Particle p1 = AddParticleTarget(owner, "pirate_raiseMorale_cas.troy", unit, 1);
+            Particle p2 = AddParticleTarget(owner, "pirate_raiseMorale_mis.troy", unit, 1);
+            Particle p3 = AddParticleTarget(owner, "pirate_raiseMorale_tar.troy", unit, 1);
```
Comment (danil179) [inline Buffs/GangplankE/GangplankE.cs]: Don't save variables that you discard. (No need `Particle p1=` part)

File: Buffs/GangplankE/GangplankE.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        public void OnDeactivate(IObjAIBase unit)
+        {
+            unit.RemoveStatModifier(_statMod);
+            
+            RemoveBuffHudVisual(_hudvisual);
+            
+            RemoveParticle(p1);
+            RemoveParticle(p2);
+            RemoveParticle(p3);
```
Comment (danil179) [inline Buffs/GangplankE/GangplankE.cs:35]: Is that even compiles?

File: Champions/Gangplank/W.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public void OnStartCasting(IChampion owner, ISpell spell, IAttackableUnit target)
+        {	
+        }
+
+        public void OnFinishCasting(IChampion owner, ISpell spell, IAttackableUnit target)
+        {	
+            float ap = owner.Stats.AbilityPower.Total * 0.1f;
+            owner.Stats.CurrentHealth += 15*ap;
```
Comment (danil179) [inline Champions/Gangplank/W.cs]: What if you exceed MaxHealth?

**Review by danil179 — CHANGES_REQUESTED** [review]
I'm not sure if that code even compiles. I don't have time to check that.

**Comment by MingSaber976** [comment]
About GangplankW, that code won't over Maximum health,tested before.

**Comment by MingSaber976** [comment]
You didn't removeParticle(if in game it still exist)
Try make "private Particle" to show particles and also can be remove.

**Comment by MingSaber976** [comment]
And in buffGameScript, you don't have "owner", try get "owner" from spell file, or just use unit(if it won't make game crash)

**Comment by xTeJk** [comment]
I do not know if these conditions for checking health points are correct, so please correct the code.

> And in buffGameScript, you don't have "owner", try get "owner" from spell file, or just use unit(if it won't make game crash)

public void OnActivate(IObjAIBase unit, ISpell ownerSpell)

ownerSpell isnt a owner definition in buffgamescript?

**Comment by MingSaber976** [comment]
ownerSpell is Spell class
So you need the "IChampion" class to fit "owner"

**Comment by MingSaber976** [comment]
Health point codes looks fine,but the middle one should be remove(because it didn't do the things)

File: Buffs/GangplankE/GangplankE.cs (hunk 2 above)
Comment (xTeJk) [reply Buffs/GangplankE/GangplankE.cs:35]: Yes, its fine, based on other skills.

File: Buffs/GangplankE/GangplankE.cs (hunk 1 above)
Comment (xTeJk) [reply Buffs/GangplankE/GangplankE.cs]: Changed

File: Champions/Gangplank/W.cs (hunk 3 above)
Comment (xTeJk) [reply Champions/Gangplank/W.cs]: Added conditions

**Review by danil179 — APPROVED** [review]
Approved. Seems to be in the right direction. Maybe it doesn't compile, but we doing incremental development.

**Pull request merged**